<div style="border:solid Chocolate 2px; padding: 40px">

<b> Кристина, привет!👋</b>

Меня зовут Алексей Гриб, и я буду ревьюером твоего проекта. 

Сразу хочу предложить в дальнейшем общаться на "ты" - надеюсь, так будет комфортнее:) Но если это неудобно, обязательно дай знать, и мы придумаем что-нибудь ещё!
    
Цель ревью - не искать ошибки в твоём проекте, а помочь тебе сделать твою работу ещё лучше, устранив недочёты и приблизив её к реальным задачам аналитика. Поэтому не расстраивайся, если что-то не получилось с первого раза - это нормально, и это поможет тебе вырасти!
    
Ты можешь найти мои комментарии, обозначенные <font color='green'>зеленым</font>, <font color='gold'>желтым</font> и <font color='red'>красным</font> цветами, например:

<br/>

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> похвала, рекомендации «со звёздочкой», полезные лайфхаки, которые сделают и без того красивое решение ещё более элегантным.
</div>

<br/>

<div class="alert alert-warning">
    <h2> Комментарий ревьюера <a class="tocSkip"> </h2>
    
<b>Некоторые замечания и рекомендации⚠️:</b> некритичные ошибки или развивающие рекомендации на будущее. 
</div>


<br/>
<div class="alert alert-block alert-danger">
<h2> Комментарий ревьюера <a class="tocSkip"></h2>

    
<b>На доработку❌:</b>
Критичные ошибки, которые обязательно нужно исправить.
</div>

Я не смогу принять проект, если в нём будет хотя бы одна критичная ошибка или несколько некритичных ошибок - тогда проект нужно будет немного доработать. Но это нестрашно - я обязательно дам тебе подсказку или укажу верное направление.
    
Пожалуйста, не удаляй мои комментарии, они будут особенно полезны для нашей работы в случае повторной проверки проекта. 
    
Ты также можешь задавать свои вопросы, реагировать на мои комментарии, делать пометки и пояснения - полная творческая свобода! Но маленькая просьба - пускай они будут отличаться от моих комментариев, это поможет избежать путаницы в нашем общении:)
Например, вот так:
    
<br/>

<div class="alert alert-info">
<h2> Комментарий студента <a class="tocSkip"> </h2>

*твой текст*
</div>
    
Давай посмотрим на твой проект!

<div class="alert alert-info">
<h2> Комментарий студента <a class="tocSkip"> </h2>

Привет, Алексей) Спасибо за проверку и ободряющий настрой), приступаю к доработке
</div>

<div style="border:solid Chocolate 2px; padding: 40px">

<b> Комментарий ревьюера v.2 </b>

Привет, Кристина! Рад снова видеть тебя и твою работу:) Давай посмотрим, что получилось на этой итерации!

<div class="alert alert-info">
<h2> Комментарий студента <a class="tocSkip"> </h2>

Алексей, привет) спасибо за быструю проверку, а то дедлайн уже в затылок дышит)))
</div>

<div style="border:solid Chocolate 2px; padding: 40px">

<b> Комментарий ревьюера v.3 </b>

Привет, Кристина! Думаю, мы уже заканчиваем - уверен, итерация последняя:) Давай смотреть!

# SQL

Во время короновируса люди меньше передвигались и появилось больше времени для чтения. Не у всех есть дома богатые библиотеки, а у некоторых и нет места на эту библиотеку,а потом и новые книги постоянно появляются, не всегда есть возможность прочитать то, что хочешь. Но к счастью появились приложения для чтения. Это заметили стартаперы — и бросились создавать приложения для тех, кто любит читать.

Задача — проанализировать базу данных крупного сервиса для чтения книг по подписке.
У нас имеется информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. 

### Описание данных

**Таблица `books`**

Содержит данные о книгах:

- `book_id` — идентификатор книги;
- `author_id` — идентификатор автора;
- `title` — название книги;
- `num_pages` — количество страниц;
- `publication_date` — дата публикации книги;
- `publisher_id` — идентификатор издателя.

**Таблица `authors`**

Содержит данные об авторах:

- `author_id` — идентификатор автора;
- `author` — имя автора.

**Таблица `publishers`**

Содержит данные об издательствах:

- `publisher_id` — идентификатор издательства;
- `publisher` — название издательства;

**Таблица `ratings`**

Содержит данные о пользовательских оценках книг:

- `rating_id` — идентификатор оценки;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, оставившего оценку;
- `rating` — оценка книги.

**Таблица `reviews`**

Содержит данные о пользовательских обзорах на книги:

- `review_id` — идентификатор обзора;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, написавшего обзор;
- `text` — текст обзора.

## План работы

1. Цель исследования
2. Подключение к базе данных и необходимых библиотек  
3. Ознакомимся с данными таблиц   
4. Посчитаем, сколько книг вышло после 1 января 2000 года
5. Для каждой книги посчитаем количество обзоров и среднюю оценку
6. Определим издательство, которое выпустило наибольшее число книг толще 50 страниц 
7. Определим автора с самой высокой средней оценкой книг — учитываем только книги с 50 и более оценками
8. Посчитаем среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> Прекрасное вступление!
    
В нём есть всё, что необходимо, чтобы понять суть проекта с первых строк отчёта!

<div class="alert alert-warning">
    <h2> Комментарий ревьюера <a class="tocSkip"> </h2>
    
<b>Некоторые замечания и рекомендации⚠️:</b> 

Твой проект станет намного удобней в использовании, если ты добавишь интерактивное оглавление — с помощью гиперссылок или в плагине toc. Материал об этом есть в блоке курса Полезные инструкции для учёбы - Оформление проекта - Изменения в Jupyter: автоматическое оглавление и горячие клавиши. Либо ты можешь использовать гиперссылки - подробнее об их применении можно прочитать тут: https://sebastianraschka.com/Articles/2014_ipython_internal_links.html.
        
</div>

<div class="alert alert-info">
<h2> Комментарий студента <a class="tocSkip"> </h2>

попробую еще раз..., я уже как-то сидела над этим))) спасибо
</div>

<div class="alert alert-success">
<h2> Комментарий ревьюера v.2 <a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> Учтено!

In [1]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import create_engine
from IPython.display import display

In [2]:
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
 db_config['pwd'],
 db_config['host'],
 db_config['port'],
 db_config['db'])
# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'})

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> Библиотеки импортировали, коннектор сделали - отлично!
    
Спасибо, что перезапускаешь ядро проекта перед отправкой на ревью - это важный шаг, который поможет избежать простых и досадных ошибок при воспроизведении решнеия:)
</div>

### Ознакомимся с данными таблиц

In [3]:
query1 = """
  SELECT*
  from books
  limit5;
  """

In [4]:
pd.io.sql.read_sql(query1, con = engine) 

book_id  author_id                                              title  \
0          1        546                                       'Salem's Lot   
1          2        465                 1 000 Places to See Before You Die   
2          3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3          4         82  1491: New Revelations of the Americas Before C...   
4          5        125                                               1776   
..       ...        ...                                                ...   
995      996        571           Wyrd Sisters (Discworld  #6; Witches #2)   
996      997        454                        Xenocide (Ender's Saga  #3)   
997      998        201                                    Year of Wonders   
998      999         94                        You Suck (A Love Story  #2)   
999     1000        509  Zen and the Art of Motorcycle Maintenance: An ...   

     num_pages publication_date  publisher_id  
0          594       2005-11-01            93  
1          992       2003-05-22           336  
2          322       2010-12-21           135  
3          541       2006-10-10           309  
4          386       2006-07-04           268  
..         ...              ...           ...  
995        265       2001-02-06           147  
996        592       1996-07-15           297  
997        358       2002-04-30           212  
998        328       2007-01-16           331  
999        540       2006-04-25           143  

[1000 rows x 6 columns]

Перед нами база данных с названиями книг, кол-м страниц, даты публикации, номер издательства и автора.

<div class="alert alert-warning">
    <h2> Комментарий ревьюера <a class="tocSkip"> </h2>
    
<b>Некоторые замечания и рекомендации⚠️:</b> 

Структурно и логически было бы более правильно сначала посмотреть на все таблицы, а уже потом приступать к проекту. Давай попробуем лучше организовать структуру проекта?
        
Высший пилотаж в исследовании таблиц - организация этого проецсса в цикле, что поможет избежать дублирования кода. Подставлять название таблицы в строку-запрос можно с помощью `f-строки` или метода `format()`. Саму строку-запрос можно создавать внутри цикла на каждой итерации.
        
</div>

<div class="alert alert-info">
<h2> Комментарий студента <a class="tocSkip"> </h2>

с ходу не получилась ф-ция, это что-то нерабочее... , не видит он таблицы, через переменную тоже не полуилось, возможно не правильно туда их передала(((     
    
    
</div>
    

<div class="alert alert-warning">
    <h2> Комментарий ревьюера v.2 <a class="tocSkip"> </h2>
    
<b>Некоторые замечания и рекомендации⚠️:</b> 

Да, ты здесь не совсем верно работаешь с циклом. `range` - это генератор последовательности цифр, он тут не совсем подходит. Вместо `range` можно использовать любую итерируемую структуру данных, включая список. 
        
Также ты не выполняешь форматирование строки, чтобы передать итерируемый объект в строку-запрос для последующего исполнения.

Код в следующей реализации должен работать:)
        
        
    for table in ['table_1', 'table_2']:
        
        query = f'SELECT * FROM {table}' -- тут форматирование через f-строку
        query = 'SELECT * FROM {}'.format(table) -- тут стандартное форматирование
</div>

<div class="alert alert-info">
<h2> Комментарий студента 2 <a class="tocSkip"> </h2>
 вот так смогла доделать, но мне кажется, что не очень удобен вывод инфо для изучения...   
возможно можно его еще доработать...)
    
    
    
</div>
    

<div class="alert alert-success">
<h2> Комментарий ревьюера v.3 <a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> Сделать удобнее поможет модуль `display`:) Я чуть улучшил твой блок анализа, посмотри, пожалуйста, на это решение:)
</div>

In [5]:
for table in ['books','reviews','authors','ratings', 'publishers']:
    query= f'SELECT * FROM {table} LIMIT 5'
    info = pd.io.sql.read_sql(query, con = engine)
    
    print(f'Данные таблицы {table}:')
    display(info.head(5))

Данные таблицы books:


book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

Данные таблицы reviews:


review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

Данные таблицы authors:


author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

Данные таблицы ratings:


rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

Данные таблицы publishers:


publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

<div class="alert alert-info">
<h2> Комментарий студента <a class="tocSkip"> </h2>

открыла вручную((    
    
    
</div>
    

<div class="alert alert-warning">
    <h2> Комментарий ревьюера v.2 <a class="tocSkip"> </h2>
    
<b>Некоторые замечания и рекомендации⚠️:</b> 

Оставил комментарий выше:)
</div>

<div class="alert alert-success">
<h2> Комментарий ревьюера v.3 <a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> Вопрос закрыт:)
</div>

In [6]:
query11 = """
  SELECT*
  from ratings
  limit5;
  """

In [7]:
pd.io.sql.read_sql(query11, con = engine) 

rating_id  book_id        username  rating
0             1        1      ryanfranco       4
1             2        1   grantpatricia       2
2             3        1    brandtandrea       5
3             4        2        lorichen       3
4             5        2     mariokeller       2
...         ...      ...             ...     ...
6451       6452     1000  carolrodriguez       4
6452       6453     1000         wendy18       4
6453       6454     1000      jarvispaul       5
6454       6455     1000           zross       2
6455       6456     1000         fharris       5

[6456 rows x 4 columns]

Здесь мы хранится инфо о рейтинге, читателе, который оставил его, уникальный номер книги, уникальный номер рейтинга.

In [8]:
query111 = """
  SELECT*
  from authors
  limit5;
  """

In [9]:
pd.io.sql.read_sql(query111, con = engine) 

author_id                                author
0            1                            A.S. Byatt
1            2        Aesop/Laura Harris/Laura Gibbs
2            3                       Agatha Christie
3            4                         Alan Brennert
4            5              Alan Moore/David   Lloyd
..         ...                                   ...
631        632         William Strunk Jr./E.B. White
632        633                           Zadie Smith
633        634                 Zilpha Keatley Snyder
634        635                    Zora Neale Hurston
635        636  Åsne Seierstad/Ingrid Christopherson

[636 rows x 2 columns]

Здесь только фио автора и его уникалный номер

In [10]:
query1111 = """
  SELECT*
  from reviews
  limit5;
  """

In [11]:
pd.io.sql.read_sql(query1111, con = engine) 

review_id  book_id        username  \
0             1        1    brandtandrea   
1             2        1      ryanfranco   
2             3        2        lorichen   
3             4        3   johnsonamanda   
4             5        3     scotttamara   
...         ...      ...             ...   
2788       2789      999      martinadam   
2789       2790     1000         wknight   
2790       2791     1000  carolrodriguez   
2791       2792     1000         wendy18   
2792       2793     1000      jarvispaul   

                                                   text  
0     Mention society tell send professor analysis. ...  
1     Foot glass pretty audience hit themselves. Amo...  
2     Listen treat keep worry. Miss husband tax but ...  
3     Finally month interesting blue could nature cu...  
4     Nation purpose heavy give wait song will. List...  
...                                                 ...  
2788  Later hospital turn easy community. Fact same ...  
2789   Change lose answer close pressure. Spend so now.  
2790  Authority go who television entire hair guy po...  
2791  Or western offer wonder ask. More hear phone f...  
2792  Republican staff bit eat material measure plan...  

[2793 rows x 4 columns]

тут у нас ники читателей, сам текст отзыва, айди книги и отзыва

In [12]:
query11111 = """
  SELECT*
  from publishers
  limit5;
  """

In [13]:
pd.io.sql.read_sql(query11111, con = engine) 

publisher_id                          publisher
0               1                                Ace
1               2                           Ace Book
2               3                          Ace Books
3               4                      Ace Hardcover
4               5  Addison Wesley Publishing Company
..            ...                                ...
335           336         Workman Publishing Company
336           337                         Wyatt Book
337           338              Yale University Press
338           339                           Yearling
339           340                     Yearling Books

[340 rows x 2 columns]

Название издат дома и его айди.

### Посчитаем, сколько книг вышло после 1 января 2000 года;

In [14]:
query2 = """  SELECT COUNT (DISTINCT(book_id))
  FROM books
  WHERE publication_date> '2000-01-01';
  """

In [15]:
pd.io.sql.read_sql(query2, con = engine) 

count
0    819

После 1 января 20 года было выпущено 819 книг.

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> Правильно!
    
Отмечу, что условие фильтрации `publication_date >= '2000-01-01' ` также было бы правильным, хоть и привело бы к другому результату - в реальных проектах о том, включается ли дата в диапазон или нет, всегда стоит уточнять заранее:)

</div>

### Для каждой книги посчитаем количество обзоров и среднюю оценку

Изучим данные в таблице `ratings`

In [16]:
query3 = """
  SELECT *
 FROM ratings
 limit 5;
  """

In [17]:
pd.io.sql.read_sql(query3, con = engine) 

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

In [18]:
query3 = """
  SELECT *
 FROM reviews
 limit 5;
  """

In [19]:
pd.io.sql.read_sql(query3, con = engine) 

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

<div class="alert alert-info">
<h2> Комментарий студента <a class="tocSkip"> </h2>

оставляю неверный вариант для истории
</div>
    

<div class="alert alert-success">
<h2> Комментарий ревьюера v.2 <a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> Хорошо, но на самом деле совсем необязательно:)

</div>

In [20]:
query3 = """
  SELECT  b.book_id,
         COUNT(r.review_id),
         AVG(rt.rating)
  FROM books AS b
  left JOIN reviews AS r ON b.book_id = r.book_id
  left JOIN ratings AS rt ON b.book_id = rt.book_id
  group by b.book_id
  order by b.book_id;
  """

In [21]:
pd.io.sql.read_sql(query3, con = engine) 

book_id  count       avg
0          1      6  3.666667
1          2      2  2.500000
2          3      9  4.666667
3          4      4  4.500000
4          5     24  4.000000
..       ...    ...       ...
995      996      9  3.666667
996      997     15  3.400000
997      998     20  3.200000
998      999      4  4.500000
999     1000     24  3.833333

[1000 rows x 3 columns]

<div class="alert alert-block alert-danger">
<h2> Комментарий ревьюера <a class="tocSkip"></h2>
    

<b>На доработку❌:</b>
Ответ не совсем верный - получилось многовато оценок. Обрати, пожалуйста, внимание, что для фильма `Twilight (Twilight  #1)` у тебя 1120 оценок. Сравни эту цифру с исходными данными в таблице `reviews` - будет ли количество оценок таким же? Думаю, сделав эту проверку, ты быстро поймёшь, в чём дело, но если нет - дай мне знать, и я помогу разобраться. 

</div>

<div class="alert alert-info">
<h2> Комментарий студента <a class="tocSkip"> </h2>

исправление 1
</div>
    

<div class="alert alert-info">
<h2> Комментарий студента <a class="tocSkip"> </h2>

исправление 2, да спасибо, проверила кол-во книг, надеюсь, сейчас все верно)
    
</div>
    

In [22]:
query_3 = """
  WITH 
  b AS (SELECT title,
               book_id
        from books),
  i AS (SELECT book_id,
               COUNT(review_id) AS review
     FROM reviews
     GROUP BY book_id),
  f AS (SELECT book_id,            
        FLOOR (AVG(rating)) AS avg_rating
        FROM ratings
        GROUP BY book_id)
   SELECT b.title,
          i.review,
          f.avg_rating
   FROM b
   left JOIN i on b.book_id = i.book_id
   left JOIN f on b.book_id = f.book_id   
   order by b.book_id;
  """

In [23]:
pd.io.sql.read_sql(query_3, con = engine) 

title  review  avg_rating
0                                         'Salem's Lot     2.0         3.0
1                   1 000 Places to See Before You Die     1.0         2.0
2    13 Little Blue Envelopes (Little Blue Envelope...     3.0         4.0
3    1491: New Revelations of the Americas Before C...     2.0         4.0
4                                                 1776     4.0         4.0
..                                                 ...     ...         ...
995           Wyrd Sisters (Discworld  #6; Witches #2)     3.0         3.0
996                        Xenocide (Ender's Saga  #3)     3.0         3.0
997                                    Year of Wonders     4.0         3.0
998                        You Suck (A Love Story  #2)     2.0         4.0
999  Zen and the Art of Motorcycle Maintenance: An ...     4.0         3.0

[1000 rows x 3 columns]

<br/>
<div class="alert alert-block alert-danger">
<h2> Комментарий ревьюера v.2 <a class="tocSkip"></h2>

    
<b>На доработку❌:</b>
Кое-что упустили: в твоём запросе 994 строки. 1 строка соответствует 1 книге. В задании сказано, что средний рейтинг и количество оценок нужно посчитать для всех книг. Давай проверим, сколько книг находится в исходной таблице `books`? Ничего ли не потерялось?
    
Также вместо `book_id` лучше вывести `title`.
</div>

<div class="alert alert-success">
<h2> Комментарий ревьюера v.3 <a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> Учтено! Ты молодец:)

</div>

<div class="alert alert-warning">
    <h2> Комментарий ревьюера v.3 <a class="tocSkip"> </h2>
    
<b>Некоторые замечания и рекомендации⚠️:</b> 
Здесь также рекомендую попробовать решить задачу без подзапросов:)
        </div>

### Определим издательство, которое выпустило наибольшее число книг толще 50 страниц — так мы исключим из анализа брошюры;

In [24]:
query4 = """
  SELECT p.publisher,
        COUNT(b.book_id)
  
  FROM books AS b
  JOIN publishers AS p ON b.publisher_id = p.publisher_id
  WHERE b.num_pages > 50
  GROUP BY p.publisher
  ORDER BY (count) desc
  LIMIT 5;
  """

In [25]:
pd.io.sql.read_sql(query4, con = engine) 

publisher  count
0             Penguin Books     42
1                   Vintage     31
2  Grand Central Publishing     25
3          Penguin Classics     24
4                    Bantam     19

Максимальное кол-во книг с кол-м страниц болше 50 было выпущено издательством `Penguin Books`, 42 книги.

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> Издательство определили верно! Спасибо, что дополнительно вывела количество книг издательства!

</div>

<div class="alert alert-warning">
    <h2> Комментарий ревьюера <a class="tocSkip"> </h2>
    
<b>Некоторые замечания и рекомендации⚠️:</b> 
В задании сказано, что нужно найти одно издательство, поэтому было бы логичнее оставить в выовде только одну строку:)
        </div>

<div class="alert alert-info">
<h2> Комментарий студента <a class="tocSkip"> </h2>

Да, Алексей) сначала так и сделала, а потом решила проверить не совпадают ли значения у следующих издательств, так потом и оставила))
</div>

<div class="alert alert-success">
<h2> Комментарий ревьюера v.2 <a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> Так оно обычно и случается:) Окей!

</div>

### Определим автора с самой высокой средней оценкой книг — учитываем только книги с 50 и более оценками

<div class="alert alert-info">
<h2> Комментарий студента <a class="tocSkip"> </h2>

 Алексей, я оставила этот вариант неправильный для истории.
</div>

<div class="alert alert-success">
<h2> Комментарий ревьюера v.2 <a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> Хорошо:)

</div>

In [26]:
query5 = """
  SELECT i.author,
         i.avg_rating,
         i.title
  FROM(
  SELECT a.author,
        AVG(r.rating) as avg_rating,
        
        COUNT(r.rating_id),
        b.title
  
  FROM books AS b
  JOIN authors AS a ON b.author_id = a.author_id
  JOIN ratings AS r ON b.book_id = r.book_id
  group by b.title, a.author_id
  HAVING COUNT(r.rating_id)>50
  order by (avg_rating) desc
  LIMIT 1) AS i;
  """

In [27]:
pd.io.sql.read_sql(query5, con = engine) 

author  avg_rating  \
0  J.K. Rowling/Mary GrandPré    4.414634   

                                               title  
0  Harry Potter and the Prisoner of Azkaban (Harr...

Самую высокую оценку набрал J.K. Rowling/Mary GrandPré.

<br/>
<div class="alert alert-block alert-danger">
<h2> Комментарий ревьюера <a class="tocSkip"></h2>
    

<b>На доработку❌:</b>
Автор с самым высоким средним рейтингом определён верно, но сам рейтинг посчитан неверно:
    
1. По условию проекта мы учитываем только книги с 50 и более оценками - в твоём запросе мы учитываем книги, у которых строго больше 50 оценок. Важно верно подбирать операторы сравнения:)
2. При исправлении первой ошибки автор у тебя поменяется, и ответ выйдет неверный - дело в том, что твой фильтр `HAVING` призван отфильтровать книги с оценками 50 и более, а данные ты группируешь по автору. В итоге фильтр фильтрует не книги, а авторов, и в такой группировке почти у каждого атвора будет 50 и более оценок, поэтому рейтинг будет рассчитан неверно - в него попадут лишние книги, которые не проходят по условию с количеством оценок.
    
Попробуй переделать запро с учётом этих рекомендаций:)

</div>

<div class="alert alert-info">
<h2> Комментарий студента <a class="tocSkip"> </h2>

Исправление 1. это какой-то мозговой лабиринт....
</div>

<div class="alert alert-success">
<h2> Комментарий ревьюера v.2 <a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> Да, задание хорошо разминает мозг:)

</div>

In [28]:
query_5 = """
  WITH         
  b AS (SELECT book_id, author_id, title
           FROM books),
         
  r AS (SELECT rating_id, rating, book_id
         FROM ratings),
         
  a AS (SELECT author_id, author
         FROM authors)      
  
  SELECT author, AVG(i.avg_rating)
  FROM (SELECT r.book_id, AVG(r.rating) AS avg_rating, b.author_id
        FROM r
        JOIN b ON r.book_id = b.book_id
        GROUP BY r.book_id, b.author_id
        HAVING COUNT(r.rating_id) >= 50) AS i     
  JOIN a ON i.author_id = a.author_id    
  GROUP BY author
  ORDER BY AVG(i.avg_rating) DESC
  LIMIT 1;
  """

In [29]:
pd.io.sql.read_sql(query_5, con = engine) 

author       avg
0  J.K. Rowling/Mary GrandPré  4.283844

<div class="alert alert-success">
<h2> Комментарий ревьюера v.2 <a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> Теперь всё верно!

</div>

<div class="alert alert-warning">
    <h2> Комментарий ревьюера v.2 <a class="tocSkip"> </h2>
    
<b>Некоторые замечания и рекомендации⚠️:</b> 
Немного об оптимизации: подзапросы - штука довольно дорогая с точки зрения вычислений, присоединения куда быстрее происходят. В данном случае твои подзапросы `b`, `r` и `a` аналогичны тому, что ты просто работаешь с исходными таблицами, поэтому они совсем не нужны. В качестве маленького упражнения на оптимизацию можешь попробовать сделать это задание без этих трёх поздапросов - тут вполне можно обойтись просто таблицами:)
        </div>

<div class="alert alert-info">
<h2> Комментарий студента v2 <a class="tocSkip"> </h2>

однозначно еще посижу над этой задачкой, я осознавла, что это далеко не оптималный вариант, но пока так получилось 
</div>

<div class="alert alert-success">
<h2> Комментарий ревьюера v.3 <a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> Да, окей, это всего лишь рекомендация, но рекомендую всё же попробовать:)

</div>

### Посчитаем среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

In [30]:
query6 = """

  WITH         
  rt AS (SELECT rating_id, username
         FROM ratings),
         
  rw AS (SELECT review_id, username
         FROM reviews)      
  
  SELECT FLOOR(AVG(i.revi))
  FROM (SELECT COUNT(rw.review_id) AS revi
        FROM rw
        WHERE 
           rw.username IN (SELECT username
                        FROM rt
                        GROUP BY username 
                        HAVING COUNT(rt.rating_id) > 50)
        GROUP BY rw.username) AS i;
  """

In [31]:
pd.io.sql.read_sql(query6, con = engine) 

floor
0   24.0

В среднем пользователи, которые поставили 50 оценок, оставляют 24 обзора.

<br/>

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> Среднее количество обзоров определили верно!

</div>

<div class="alert alert-warning">
    <h2> Комментарий ревьюера <a class="tocSkip"> </h2>
    
<b>Некоторые замечания и рекомендации⚠️:</b> 
Говоря про отзывы, было бы логичнее округлить ответ до целого значения в меньшую сторону - отзыв может быть или целым, или его вообще нет:)
        </div>

<br/>

<div class="alert alert-success">
<h2> Комментарий ревьюера v.2 <a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> Есть!

</div>

<div class="alert alert-warning">
    <h2> Комментарий ревьюера <a class="tocSkip"> </h2>
    
<b>Некоторые замечания и рекомендации⚠️:</b> 
По завершении проекта стоит сделать итоговый вывод о проведённой работе: можно тезисно перечислить основные шаги, а также привести ответы на полученные вопросы.
        </div>

<div class="alert alert-info">
<h2> Комментарий студента <a class="tocSkip"> </h2>

добавляю вывод.
</div>

<br/>

<div class="alert alert-success">
<h2> Комментарий ревьюера v.2 <a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> Есть!

</div>

## Вывод  
  Мы работали с 5 таблицами, просмотрели каждую. Узнали, что , начиная с 01.01.20 г ,было выпущено 819 книг, самое активное издательство, издающее книги с 50 стр и более, это Penguin Books. Джоан Роулинг (J.K. Rowling/Mary GrandPré) - автор с наивысшим средним рейтингом среди самых популярных авторов. Оказываются есть пользователи, которые остаивли более 50 оценок и в среднем они оставляют 24 обзора.

<div style="border:solid Chocolate 2px; padding: 40px">

# Комментарий ревьюера: общий вывод по проекту.

Кристина, отлично поработала над проектом - почти все задания у тебя выполнены верно!

Мне нравится твой аналитический подход к выполнению проекта, ты соблюдаешь структуру работы, выполняешь её последовательно - это очень хорошо! Шаги проекта выполнены по порядку согласно плану проекта, нет смысловых и структурных ям. Важно, что не забываешь про выводы.
    
SQL запросы написаны хорошо и оптимально, в них нет избыточных действий - это здорово, кажется, ты неплохо разобралась в теме!
    
В проекте есть задания, к которым стоит вернуться - их я отметил комментариями красного цвета, в которых оставил для тебя подсказки о верном направлении. Важно исправить эти задания, чтобы проект был принят. 
    
Также в работе есть небольшие рекомендации по улучшению проекта - ответы на вопросы, выводы и прочая мелочь, которая не займёт много времени, но сделает твой проект стилистически более совершенным.
    
Если о том, что нужно сделать в рамках комментариев, будут возникать вопросы - оставь их, пожалуйста, в комментариях, и я отвечу на них во время следующего ревью.
    
Также буду рад ответить на любые твои вопросы по проекту или на какие-либо другие, если они у тебя имеются - оставь их в комментариях, и я постараюсь ответить:)
    
Жду твой проект на повторном ревью. До встречи:)

<div style="border:solid Chocolate 2px; padding: 40px">

# Комментарий ревьюера: общий вывод по проекту v.2 .

Кристина, спасибо, что прислала проект на повторное ревью. Ты учла все мои комментарии - спасибо за ответственный подход к выполнению проекта:)
    
Одно задание всё ещё не удалось решить верно - это задание про метрики для книги. Я оставил тебе комментарий с новой подсказкой:)
    
Жду твою работу на повторной проверке!

<div style="border:solid Chocolate 2px; padding: 40px">

# Комментарий ревьюера: общий вывод по проекту v.3.

Кристина, теперь все задания проекта решены верно - принимаю твой проект!
    
Спасибо за качественную и упорную работу над проектом, желаю успешного завершения обучения и лёгкого трудоустройства:)